### PR TITLE
Add explicit explications/prints for bunches in the nilearn examples

### PR DESCRIPTION
**Pull request by banilo**
Newcommers to nilearn (not seldomly: and to python) have expressed difficulty in understanding the format in which the datasets are represented.

It would therefore be helpful to
- show in the code that bunches are minor modifications of dictionaries
- print where the (fetched/located) files are actually located on the disk
- show how to access the important dataset parts right after first access

### CODE REVIEW

**Comment by lesteve** [comment]
The description is rather verbose generally so I am not sure it helps printing that much information when running the example.

I'd be in favour of updating the documentation [here](http://nilearn.github.io/building_blocks/manipulating_mr_images.html#datasets) to at the very least mention the "description" field.

**Comment by banilo** [comment]
I honestly believe that (a little more) output related to the Bunches would help a lot of users Looking at the ones first examples, the Bunches (instead of an ordinary dictionary) are a major hurdle. In particular, for users that may be new to both Python and nilearn.

Autobiographical note: I found the fetch/bunches mechanisms wierd myself, not long ago. Now it appears obvious.

**Comment by banilo** [comment]
Compromise:
- leave out the description print (I agree it is verbose)
- leave in the printing of the filenames -> makes clear that there are actual files, with actual locations, that are the result of an actual download

**Comment by AlexandreAbraham** [comment]
:+1:

**Comment by lesteve** [comment]
Why not have a separate example like "how to use nilearn datasets fetcher" rather than clogging existing ones?

**Comment by AlexandreAbraham** [comment]
Because a lot of people asked for the specific path to the datasets after running examples. I tried to underline that in the dataset fetcher but it may not be enough.

**Comment by banilo** [comment]
..for pragmatic reasons. If a new user downloads nilearn and wants to use
to it for example specifically to do a "searchlight". He/she will mostly
likely find the example on searchlight at some point. At the first lines
there is the bunch thing. Most likely that person will not have the hunch
"oh, perhaps there is another example relevant to me right now, that I
should look at first, because there is some special structure for datasets".

That's why those few lines that print the data decomplexify the first
contact.



2015-04-08 10:52 GMT+02:00 Loïc Estève <notifications@github.com>:

> Why not have a separate example like "how to use nilearn datasets fetcher"
> rather than clogging existing ones?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/nilearn/nilearn/pull/531#issuecomment-90847315>.
>



-- 
Viele Grüße,
Danilo

**Comment by lesteve** [comment]
Fair enough, does that mean that we want to add these kind of lines to every single example going forward?

**Comment by banilo** [comment]
That would be my suggestions, yes (print only the data file paths) - as now
in the PR.

We can cut the description, however, agreed.

2015-04-08 11:07 GMT+02:00 Loïc Estève <notifications@github.com>:

> Fair enough, does that mean that we want to add these kind of lines to
> every single example going forward?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/nilearn/nilearn/pull/531#issuecomment-90854965>.
>



-- 
Viele Grüße,
Danilo

**Comment by lesteve** [comment]
OK, remove the print(description) squash your commits and this is good to go.

Still think it'd be great to tackle my earlier comment in a separate PR while you are on this topic:
> I'd be in favour of updating the documentation [here](http://nilearn.github.io/building_blocks/manipulating_mr_images.html#datasets) to at the very least mention the "description" field.

**Comment by banilo** [comment]
+1

Yes, I will do both this afternoon.

2015-04-08 11:31 GMT+02:00 Loïc Estève <notifications@github.com>:

> OK, remove the print(description) squash your commits and this is good to
> go.
>
> Still think it'd be great to tackle my earlier comment in a separate PR
> while you are on this topic:
>
> I'd be in favour of updating the documentation here
> <http://nilearn.github.io/building_blocks/manipulating_mr_images.html#datasets>
> to at the very least mention the "description" field.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/nilearn/nilearn/pull/531#issuecomment-90860673>.
>



-- 
Viele Grüße,
Danilo

**Comment by banilo** [comment]
Cleaned + squashed!

File: examples/connectivity/plot_canica_resting_state.py [hunk 1]
```diff
@@ -26,6 +26,9 @@
 adhd_dataset = datasets.fetch_adhd()
 func_filenames = adhd_dataset.func  # list of 4D nifti files for each subject
 
+# print basic information on the dataset
+print('Functional nifti images are at: %s' % adhd_dataset.func[0])  # 4D data
```
Comment (lesteve) [inline examples/connectivity/plot_canica_resting_state.py]: I would use "First functional nifti image is at: " otherwise this is a bit misleading.

Other examples would need to be harmonised in a similar fashion.

Comment (bthirion) [inline examples/connectivity/plot_canica_resting_state.py]: Indeed, this would give a more accurate information. 
LGTM otherwise.

Comment (banilo) [inline examples/connectivity/plot_canica_resting_state.py]: +1

**Comment by banilo** [comment]
Updated this one. I tried to put more pedagogic emphasis on the difference between a) number of nifti images and b) whether those contain one or more images.

I imagine this is particularly confusing (on top of the bunches).

File: examples/connectivity/plot_canica_resting_state.py [hunk 2]
```diff
@@ -26,6 +26,10 @@
 adhd_dataset = datasets.fetch_adhd()
 func_filenames = adhd_dataset.func  # list of 4D nifti files for each subject
 
+# print basic information on the dataset
+print('First functional nifti image (4D) is at: %s' %
+      adhd_dataset.func[0])  # 4D data
```
Comment (lesteve) [inline examples/connectivity/plot_canica_resting_state.py]: I don't think it should say first here since the list is of size 1, maybe something like 'Functional 4D image is at: '

Comment (banilo) [inline examples/connectivity/plot_canica_resting_state.py]: I hate to disagree, yet `datasets.fetch_adhd()` returns a list if 40 niftis in the 'func' dictionary entry on my machine.

Comment (lesteve) [inline examples/connectivity/plot_canica_resting_state.py]: You are right indeed this comment does apply to nyu_dataset below though.

File: examples/connectivity/plot_ica_resting_state.py [hunk 3]
```diff
@@ -13,6 +13,11 @@
 nyu_dataset = datasets.fetch_nyu_rest(n_subjects=1)
 func_filename = nyu_dataset.func[0]
 
+# print basic information on the dataset
+print('First anatomical nifti image (3D) is at: %s' % nyu_dataset.anat_anon[0])
+print('First functional nifti image (4D) is at: %s' %
+      nyu_dataset.func[0])  # 4D data
```
Comment (lesteve) [inline examples/connectivity/plot_ica_resting_state.py]: > I don't think it should say first here since the list is of size 1, maybe something like 'Functional 4D image is at: '

This is where my earlier comment was supposed to end up.

Comment (banilo) [inline examples/connectivity/plot_ica_resting_state.py]: You are right, but it depends on 'n_subjects' argument above. To also make
the datasets clearer I would therefore suggest to leave the printings as
they are and increase the #subjects to 2 instead. The aim is educational.

2015-04-13 10:36 GMT+02:00 Loïc Estève <notifications@github.com>:

> In examples/connectivity/plot_ica_resting_state.py
> <https://github.com/nilearn/nilearn/pull/531#discussion_r28221872>:
>
> > @@ -13,6 +13,11 @@
> >  nyu_dataset = datasets.fetch_nyu_rest(n_subjects=1)
> >  func_filename = nyu_dataset.func[0]
> >
> > +# print basic information on the dataset
> > +print('First anatomical nifti image (3D) is at: %s' % nyu_dataset.anat_anon[0])
> > +print('First functional nifti image (4D) is at: %s' %
> > +      nyu_dataset.func[0])  # 4D data
>
>  I don't think it should say first here since the list is of size 1,
> maybe something like 'Functional 4D image is at: '
>
> This is where my earlier comment was supposed to end up.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/nilearn/nilearn/pull/531/files#r28221872>.
>



-- 
Viele Grüße,
Danilo

Comment (lesteve) [inline examples/connectivity/plot_ica_resting_state.py]: Just write something like 'First subject functional nifti image (4D) is at: '

Comment (lesteve) [inline examples/connectivity/plot_ica_resting_state.py]: Downloading the 2nd subject dataset if you are not using it doesn't qualify as educational to me, unless you think waiting for 15 minutes without any good reason is an educational experience ;-).

**Comment by banilo** [comment]
Ok, how about now?

File: examples/decoding/plot_haxby_different_estimators.py [hunk 4]
```diff
@@ -10,7 +10,13 @@
 
 ### Fetch data using nilearn dataset fetcher ################################
 from nilearn import datasets
-haxby_dataset = datasets.fetch_haxby(n_subjects=1)
+haxby_dataset = datasets.fetch_haxby(n_subjects=2)
```
Comment (lesteve) [inline examples/decoding/plot_haxby_different_estimators.py]: Any good reason why this was changed to 2 ?

I guess you want to be consistent and use something like "First subject functional ..." in your print statement below and all the other haxby examples.

File: examples/decoding/plot_haxby_grid_search.py [hunk 5]
```diff
@@ -38,6 +38,10 @@
 import numpy as np
 haxby_dataset = datasets.fetch_haxby_simple()
 
+# print basic information on the dataset
+print('Mask nifti images are located at: ' % haxby_dataset.mask[0])
+print('Functional nifti images are located at: ' % haxby_dataset.func[0])
```
Comment (lesteve) [inline examples/decoding/plot_haxby_grid_search.py]: For, `haxby_dataset.func` is a string. Can you make sure all the examples run fine to get rid for other mistakes like this ?

**Comment by banilo** [comment]
datasets.fetch_haxby() appears to be broken for some reason (.anat is empty and .func also). Ideas?

**Comment by lesteve** [comment]
> datasets.fetch_haxby() appears to be broken for some reason (.anat is empty and .func also). Ideas?

Hmmm seems to work fine for me. I even tried deleting ~/nilearn_data/haxby2001{,_simple} to make sure that wasn't one of the usual problems that you only get when you haven't downloaded the dataset yet.

**Comment by banilo** [comment]
> seems to work fine for me

Ok, I think I found the culprit. The Bunches does not appear to be filled out correctly if the fetch_simuli option is used

`haxby_dataset = datasets.fetch_haxby(n_subjects=1, fetch_stimuli=True)`

**Comment by banilo** [comment]
.func is actually overwritten, if fetch_stimuli is used, and the imaging data is not in the output bunch anymore:

https://github.com/nilearn/nilearn/blob/master/nilearn/datasets.py#L1226

Is this the expected behavior?

**Comment by AlexandreAbraham** [comment]
No. Good catch !

**Comment by lesteve** [comment]
> No. Good catch !

created https://github.com/nilearn/nilearn/issues/544

File: examples/plot_nifti_simple.py [hunk 6]
```diff
@@ -10,7 +10,12 @@
 
 from nilearn import datasets
 from nilearn.input_data import NiftiMasker
-nyu_dataset = datasets.fetch_nyu_rest(n_subjects=1)
+nyu_dataset = datasets.fetch_nyu_rest(n_subjects=2)
```
Comment (lesteve) [inline examples/plot_nifti_simple.py]: No good reason to use 2 subjects I reckon.

**Comment by banilo** [comment]
Ok, this PR should be good now, too.

File: examples/decoding/plot_oasis_vbm.py [hunk 7]
```diff
@@ -45,13 +45,19 @@
 from nilearn import datasets
 from nilearn.input_data import NiftiMasker
 
-n_subjects = 100  # more subjects requires more memory
+n_subjects = 2  # more subjects requires more memory
```
Comment (lesteve) [inline examples/decoding/plot_oasis_vbm.py]: why this change?

File: examples/decoding/plot_haxby_stimuli.py [hunk 8]
```diff
@@ -12,8 +12,14 @@
...
-haxby_dataset = datasets.fetch_haxby(n_subjects=0, fetch_stimuli=True)
-stimulus_information = haxby_dataset.stimuli
+haxby_dataset = datasets.fetch_haxby(n_subjects=1, fetch_stimuli=True)
+# stimulus_information = haxby_dataset.stimuli
+
+# print basic information on the dataset
+print('First subject anatomical nifti image (3D) located is at: %s' %
+      haxby_dataset.anat[0])
+print('First subject functional nifti image (4D) is located at: %s' %
+      haxby_dataset.func[0])
```
Comment (lesteve) [inline examples/decoding/plot_haxby_stimuli.py]: This example only uses the stimuli data, I don't think there is any point downloading the data for one subject.

Comment (lesteve) [inline examples/decoding/plot_haxby_stimuli.py]: Given that this example is broken, it'd be nice if you could fix it and make sure that all the other examples run.

Comment (banilo) [inline examples/decoding/plot_haxby_stimuli.py]: > Given that this example is broken, it'd be nice if you could fix it

You are the one who fixed it - it was the fetch_stimulus-issue.

> that all the other examples run.

All work until printing the new descriptions.

File: examples/decoding/plot_oasis_vbm.py (hunk 7 above)
Comment (banilo) [inline examples/decoding/plot_oasis_vbm.py]: Ok, changed it back.

**Comment by lesteve** [comment]
Looks fine, consistency in messages could be improved but let's leave that for another PR.
